### PR TITLE
fix auto detection of langchain model streamable

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12152?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12152/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12152
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/langchain/test_langchain_model_export.py [hunk 1]
```diff
@@ -3152,3 +3152,41 @@ def test_load_chain_with_model_config_overrides_saved_config(chain_model_signatu
...
+        model_info = mlflow.langchain.log_model(
+            lc_model=model,
+            artifact_path="model",
+            streamable=streamable,
+        )
+
+        if streamable is True:
+            assert model_info.flavors["langchain"]["streamable"] is True
+        else:
+            assert model_info.flavors["langchain"]["streamable"] is False
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:3192]:
```suggestion
        assert model_info.flavors["langchain"]["streamable"] is bool(streamable)
```

File: tests/langchain/test_langchain_model_export.py [hunk 2]
```diff
@@ -3152,3 +3152,41 @@ def test_load_chain_with_model_config_overrides_saved_config(chain_model_signatu
...
+        model_info = mlflow.langchain.log_model(
+            lc_model=chain,
+            artifact_path="model",
+            streamable=streamable,
+        )
+
+    if streamable is False:
+        assert model_info.flavors["langchain"]["streamable"] is False
+    else:
+        assert model_info.flavors["langchain"]["streamable"] is True
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:3177]:
```suggestion
    expected = (streamable is None) or streamable
    assert model_info.flavors["langchain"]["streamable"] is expected
```

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -347,7 +347,7 @@ def load_retriever(persist_directory):
             mlflow_model.metadata.update(schema)
 
     if streamable is None:
-        streamable = isinstance(lc_model, lc_runnables_types())
+        streamable = hasattr(lc_model, "stream")
```
Comment (harupy) [inline mlflow/langchain/__init__.py:350]: can we update the docstring?

File: tests/langchain/test_langchain_model_export.py [hunk 4]
```diff
@@ -3152,3 +3152,41 @@ def test_load_chain_with_model_config_overrides_saved_config(chain_model_signatu
...
+@pytest.mark.skipif(
+    Version(langchain.__version__) < Version("0.0.311"), reason="feature not existing"
+)
+@pytest.mark.parametrize("streamable", [True, False, None])
+def test_langchain_model_streamable_param_in_log_model(streamable, fake_chat_model):
+    from langchain.prompts import ChatPromptTemplate
+    from langchain.schema.output_parser import StrOutputParser
+
+    prompt = ChatPromptTemplate.from_template("What's your favorite {industry} company?")
+    chain = prompt | fake_chat_model | StrOutputParser()
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:3166]: Can we test a few more types?

File: mlflow/langchain/__init__.py [hunk 5]
```diff
@@ -542,8 +542,8 @@ def load_retriever(persist_directory):
             .. Note:: Experimental: This parameter may change or be removed in a future
                                     release without warning.
         streamable: A boolean value indicating if the model supports streaming prediction. If
-            True, the model must implement `stream` method. If None, the model's streamability
-            is inferred from the model type. Default to `None`.
+            True, the model must implement `stream` method. If None, we set streamable=True if
+            the model has `stream` attribute. Default to `None`.
```
Comment (harupy) [inline mlflow/langchain/__init__.py:546]: I avoid using `we` or `you` for user-facing messages.

Comment (serena-ruan) [reply mlflow/langchain/__init__.py]: Updated!

File: tests/langchain/test_langchain_model_export.py [hunk 6]
```diff
@@ -3152,3 +3140,33 @@ def test_load_chain_with_model_config_overrides_saved_config(chain_model_signatu
...
+
+
+@pytest.mark.skipif(
+    Version(langchain.__version__) < Version("0.0.311"), reason="feature not existing"
+)
+@pytest.mark.parametrize("streamable", [True, False, None])
+def test_langchain_model_streamable_param_in_log_model(streamable, fake_chat_model):
+    for model_type in lc_runnables_types():
+        with mock.patch("mlflow.langchain._save_model"), mlflow.start_run():
+            model = mock.MagicMock(spec=model_type)
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:3152]: can we use real models?

Comment (serena-ruan) [reply tests/langchain/test_langchain_model_export.py:3152]: Why?

Comment (harupy) [reply tests/langchain/test_langchain_model_export.py:3152]: Just because it works fine with mocks doesn't mean it does with real models, right?

Comment (harupy) [reply tests/langchain/test_langchain_model_export.py:3152]: We shouldn't use mocks to cut corners or write less code.

File: mlflow/langchain/__init__.py [hunk 7]
```diff
@@ -542,8 +542,8 @@ def load_retriever(persist_directory):
             .. Note:: Experimental: This parameter may change or be removed in a future
                                     release without warning.
         streamable: A boolean value indicating if the model supports streaming prediction. If
-            True, the model must implement `stream` method. If None, the model's streamability
-            is inferred from the model type. Default to `None`.
+            True, the model must implement `stream` method. If None, If None, streamable is
+            set to True if the model has the stream attribute. Default to `None`.
```
Comment (harupy) [inline mlflow/langchain/__init__.py:546]:
```suggestion
            set to True if the model implements `stream` method. Default to `None`.
```

To align with the phrasing for `True`.

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

**Pull request merged**